pytorch_lightning/utilities/auto_restart.py

```diff
@@ -70,31 +70,37 @@ def worker_id(self) -> int:
 
     def __iter__(self) -> Iterator[Any]:
         self._current_iteration = 0
+        self.i = 0
+        self.sampler_iter = iter(self._sampler)
+        return self
+
+    def __next__(self):
         # the `state dict` was cached as workers were unavailable before.
         if self._cached_state_dict is not None:
             self._load_non_random_state(self._cached_state_dict)
 
-        i = 0
-        sampler_iter = iter(self._sampler)
-        while i < self._current_iteration:
-            next(sampler_iter)
-            i += 1
+        while self.i < self._current_iteration:
+            next(self.sampler_iter)
+            self.i += 1
 
         # here: i == self._current_iteration
         if self._cached_state_dict is not None:
             self._cached_state_dict = None
 
         # recreate iterator to be sure loading is reflected there as well
-        while True:
-            self._current_iteration += 1
-            try:
-                yield next(sampler_iter)
-            except StopIteration:
-                break
+        self._current_iteration += 1
+        self.i += 1
+        has_raised = False
+        try:
+            return next(self.sampler_iter)
+        except StopIteration:
+            has_raised = True
 
         self._current_iteration = 0
         self._cached_state_dict = None
         self.restarting = False
```

Comment:
this code is meant to run in any case, so you could simply put it in the finally block and don't need the bool to re-raise

Reply:
Good catch !

Reply:
I added it back as it can hep with wrong usage from the FastForwardSampler. 

The test `test_fast_forward_on_sequential_sampler` is an example where the FastForwardSampler isn't fully consumed before calling iter again. I think it is safer to always reset entirely when calling iter

Reply:
but can the code not be simplified by

```
try:
    return next(self.sampler_iter)
except StopIteration:
    raise StopIteration
finally:
    self._current_iteration = 0
    self._cached_state_dict = None
    self.restarting = False
```
finally is guaranteed to run, so no need to jump using has_raised

Reply:
no, I'm wrong. this won't work. no simplification possible because its in the `next` method, not `iter` with yield